alvr/adb/src/commands.rs

```diff
@@ -0,0 +1,333 @@
+// https://android.googlesource.com/platform/packages/modules/adb/+/refs/heads/main/docs/user/adb.1.md
+
+use crate::parse::{self, Device, ForwardedPorts};
+use alvr_filesystem::Layout;
+use anyhow::{anyhow, Context, Result};
+use std::{collections::HashSet, io::Cursor, path::PathBuf, process::Command, time::Duration};
+use zip::ZipArchive;
+
+#[cfg(windows)]
+use std::os::windows::process::CommandExt;
+
+// https://developer.android.com/tools/releases/platform-tools#revisions
+// NOTE: At the time of writing this comment, the revisions section above
+// shows the latest version as 35.0.2, but the latest that can be downloaded
+// by specifying a version is 35.0.0
+const PLATFORM_TOOLS_VERSION: &str = "-latest"; // E.g. "_r35.0.0"
+
+#[cfg(target_os = "linux")]
+const PLATFORM_TOOLS_OS: &str = "linux";
+#[cfg(target_os = "macos")]
+const PLATFORM_TOOLS_OS: &str = "darwin";
+#[cfg(windows)]
+const PLATFORM_TOOLS_OS: &str = "windows";
+
+const REQUEST_TIMEOUT: Duration = Duration::from_secs(5);
+
+fn get_command(adb_path: &str, args: &[&str]) -> Command {
+    let mut command = Command::new(adb_path);
+    command.args(args);
+
+    #[cfg(windows)]
+    command.creation_flags(0x08000000); // CREATE_NO_WINDOW
+
+    command
+}
+
+fn download(url: &str, progress_callback: impl Fn(usize, Option<usize>)) -> Result<Vec<u8>> {
+    let agent = ureq::builder()
+        .timeout_connect(REQUEST_TIMEOUT)
+        .timeout_read(REQUEST_TIMEOUT)
+        .build();
+    let response = agent.get(url).call()?;
+    let maybe_expected_size: Option<usize> = response
+        .header("Content-Length")
+        .and_then(|l| l.parse().ok());
+    let mut result = maybe_expected_size
+        .map(Vec::with_capacity)
+        .unwrap_or_default();
+    let mut reader = response.into_reader();
+    let mut buffer = [0; 65535];
+    loop {
+        let read_count: usize = reader.read(&mut buffer)?;
+        if read_count == 0 {
+            break;
+        }
+        result.extend_from_slice(&buffer[..read_count]);
+        let current_size = result.len();
+        (progress_callback)(current_size, maybe_expected_size);
+    }
+
+    Ok(result)
+}
```

Comment:
This seems pretty manual for just downloading a file

Reply:
Well the code is from @E1int